File: Maximum Bipartite Matching.py
class Solution:
    def try_kunh(self, u, G, used, mt):
        for v in range(len(G[0])):
            if(G[u][v] and used[v] == False):
                used[v] = True
                if(mt[v] < 0 or self.try_kunh(mt[v], G, used, mt)):
                    mt[v] = u
                    return True
        return False
    def maximumMatch(self, G):
        m = len(G)
        n = len(G[0])
        N = max(n,m)
        ans = 0
        mt = [-1 for i in range(N)]
        for i in range(m):
            used = [False for j in range(N)]
            if(self.try_kunh(i, G, used, mt)):
                ans = ans+1
        return ans

File: Maximum Bipartite Matching.cpp
class Solution {
public:
	vector<int>mt;
	vector<bool>used;
	bool try_kunh(int u, vector<vector<int>>&G){
	    for (int v = 0; v < G[0].size(); v++){
	        if (G[u][v] && !used[v])
	        {
	            used[v] = true;
	            if (mt[v] < 0 || try_kunh(mt[v], G))
	            {
	                mt[v] = u;
	                return true;
	            }
	        }
	    }
    return false;
	}
	int maximumMatch(vector<vector<int>>&G){
		int m = G.size();
		int n = G[0].size();
		int N = max(n, m);
		int ans = 0;
		mt.assign(N, -1);
		for(int i = 0; i < m; i++){
			used.assign(N, false);
			if(try_kunh(i, G))
				ans++;
		}
		return ans;
	}

};
